src/deferred.rs

```diff
@@ -0,0 +1,166 @@
+use std::mem;
+use std::ptr;
+
+/// Provides methods to dispatch a call to a `FnOnce()` from a trait object.
+pub trait Callback {
+    /// Calls the function from a trait object on the stack.
+    ///
+    /// This will copy `self`, call the function, and finally drop the copy.
+    /// This method may be called only once, and `self` must not be dropped after that (tip: pass
+    /// it to `std::mem::forget`).
+    unsafe fn copy_and_call(&self);
+
+    /// Calls the function from a trait object on the heap.
+    fn call_box(self: Box<Self>);
+}
+
+impl<F: FnOnce()> Callback for F {
+    #[inline]
+    unsafe fn copy_and_call(&self) {
+        let f: Self = ptr::read(self);
+        f();
+    }
+
+    #[inline]
+    fn call_box(self: Box<Self>) {
+        let f: Self = *self;
+        f();
+    }
+}
+
+/// The representation of a trait object like `&SomeTrait`.
+///
+/// This struct has the same layout as types like `&SomeTrait` and `Box<AnotherTrait>`.
+///
+/// It is actually already provided as `std::raw::TraitObject` gated under the nightly `raw`
+/// feature. But we don't use nightly Rust, so the struct was simply copied over into Crossbeam.
+///
+/// If the layout of this struct changes in the future, Crossbeam will break, but that is a fairly
+/// unlikely scenario.
+// FIXME(stjepang): When feature `raw` gets stabilized, use `std::raw::TraitObject` instead.
+#[repr(C)]
+#[derive(Copy, Clone)]
+struct TraitObject {
+    data: *mut (),
+    vtable: *mut (),
+}
+
+/// Some space to keep a `FnOnce()` object on the stack.
+type Data = [usize; 3];
```

Comment:
`3` here is a magic number.  Maybe have a `const` for it?  Also, I would like to ask how `3` is chosen here.

Reply:
I've added a constant with some explanation.